File: xllm/pybind/llm.py
# Copyright 2025-2026 The xLLM Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/jd-opensource/xllm/blob/main/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import os
import signal
import sys
import threading
import time
import uuid
from collections.abc import Callable, Sequence
from typing import Any

from xllm_export import LLMMaster, Options, RequestOutput, RequestParams, VLMMaster

from . import utils
from .errors import ValidationError
from .params import (
    BeamSearchParams,
    PoolingParams,
    SamplingParams,
    _RequestParamsProxy,
    to_request_params,
    to_request_params_list,
)


def _get_tqdm(
    use_tqdm: bool | Callable[..., Any],
) -> Callable[..., Any] | None:
    if not use_tqdm:
        return None
    if callable(use_tqdm):
        return use_tqdm
    try:
        from tqdm import tqdm
    except ImportError as exc:
        raise ImportError(
            "tqdm is required when use_tqdm=True. Set use_tqdm=False to disable the progress bar."
        ) from exc
    return tqdm


class BeamSearchOutput:
    def __init__(self, output: RequestOutput) -> None:
        self.prompt = output.prompt
        self.sequences = output.outputs
        self.status = output.status
        self.usage = output.usage
        self.request_output = output


class EmbeddingOutputs:
    def __init__(self, output: RequestOutput) -> None:
        embedding = []
        if output.outputs and len(output.outputs) > 0:
            embedding = output.outputs[0].embeddings
        self.embedding = embedding
        self.embeddings = embedding


class EmbeddingOutput:
    def __init__(self, output: RequestOutput) -> None:
        self.prompt = output.prompt
        self.outputs = EmbeddingOutputs(output)
        self.status = output.status
        self.usage = output.usage
        self.request_output = output


class LLM:
    @staticmethod
    def _is_vllm_style_inputs(prompts: object) -> bool:
        if isinstance(prompts, dict):
            return True
        if isinstance(prompts, list) and prompts and all(isinstance(x, dict) for x in prompts):
            return True
        return False

    def __init__(
        self,
        model: str,
        task: str = "generate",
        runner: str | None = None,
        draft_model: str | None = "",
        limit_image_per_prompt: int = 8,
        block_size: int = 128,
        max_cache_size: int = 0,
        max_memory_utilization: float = 0.8,
        enable_prefix_cache: bool = True,
        max_tokens_per_batch: int = 10240,
        max_seqs_per_batch: int = 1024,
        max_tokens_per_chunk_for_prefill: int = -1,
        num_speculative_tokens: int = 0,
        speculative_algorithm: str = "MTP",
        num_request_handling_threads: int = 4,
        communication_backend: str = "hccl",
        rank_tablefile: str = "",
        expert_parallel_degree: int = 0,
        enable_chunked_prefill: bool = True,
        master_node_addr: str = "",
        instance_role: str = "DEFAULT",
        transfer_listen_port: int = 26000,
        nnodes: int = 1,
        node_rank: int = 0,
        dp_size: int = 1,
        cp_size: int = 1,
        ep_size: int = 1,
        instance_name: str = "",
        enable_disagg_pd: bool = False,
        enable_pd_ooc: bool = False,
        enable_schedule_overlap: bool = False,
        kv_cache_transfer_mode: str = "PUSH",
        disable_ttft_profiling: bool = False,
        enable_forward_interruption: bool = False,
        enable_graph: bool = False,
        enable_graph_mode_decode_no_padding: bool = False,
        enable_prefill_piecewise_graph: bool = False,
        max_tokens_for_graph_mode: int = 2048,
        enable_shm: bool = False,
        is_local: bool = True,
        input_shm_size: int = 1024,
        output_shm_size: int = 128,
        kv_cache_dtype: str = "auto",
        use_cpp_chat_template: bool = True,
        disable_log_stats: bool = True,
        enable_sleep_mode: bool = False,
        **kwargs: Any,
    ) -> None:
        signal.signal(signal.SIGTERM, lambda s, f: sys.exit(0))
        signal.signal(signal.SIGINT, lambda s, f: sys.exit(0))

        if kwargs:
            unknown = ", ".join(sorted(kwargs.keys()))
            raise TypeError(f"Unexpected keyword arguments: {unknown}")

        if runner is not None:
            if runner != "pooling":
                raise ValueError(f"unsupported runner: {runner}")
            task = "embed"

        if not os.path.exists(model):
            raise ValueError(f"model {model} not exists")

        model_type, backend = utils._infer_model_type_and_backend(model)
        if backend == "dit":
            raise ValueError("LLM does not support DiT backend models")
        if model_type is None:
            raise ValueError("model_type is required for offline inference")
        utils._configure_cpp_chat_template(use_cpp_chat_template, model_type)

        options = Options()
        options.model_path = model
        options.task_type = task
        options.draft_model_path = draft_model
        options.backend = backend
        options.limit_image_per_prompt = limit_image_per_prompt
        options.block_size = block_size
        options.max_cache_size = max_cache_size
        options.max_memory_utilization = max_memory_utilization
        options.enable_prefix_cache = enable_prefix_cache
        options.max_tokens_per_batch = max_tokens_per_batch
        options.max_seqs_per_batch = max_seqs_per_batch
        options.max_tokens_per_chunk_for_prefill = max_tokens_per_chunk_for_prefill
        options.num_speculative_tokens = num_speculative_tokens
        options.speculative_algorithm = speculative_algorithm
        options.num_request_handling_threads = num_request_handling_threads
        options.communication_backend = communication_backend
        options.rank_tablefile = rank_tablefile
        options.expert_parallel_degree = expert_parallel_degree
        options.enable_chunked_prefill = enable_chunked_prefill
        if master_node_addr:
            options.master_node_addr = master_node_addr
        else:
            free_port = utils.get_free_port()
            options.master_node_addr = "127.0.0.1:" + str(free_port)
        options.transfer_listen_port = transfer_listen_port
        options.nnodes = nnodes
        options.node_rank = node_rank
        options.dp_size = dp_size
        options.cp_size = cp_size
        options.ep_size = ep_size
        options.instance_name = instance_name
        options.enable_disagg_pd = enable_disagg_pd
        options.enable_schedule_overlap = enable_schedule_overlap
        options.enable_pd_ooc = enable_pd_ooc
        options.kv_cache_transfer_mode = kv_cache_transfer_mode
        options.disable_ttft_profiling = disable_ttft_profiling
        options.enable_forward_interruption = enable_forward_interruption
        options.enable_graph = enable_graph
        options.enable_graph_mode_decode_no_padding = enable_graph_mode_decode_no_padding
        options.enable_prefill_piecewise_graph = enable_prefill_piecewise_graph
        options.max_tokens_for_graph_mode = max_tokens_for_graph_mode
        options.enable_offline_inference = True
        options.spawn_worker_path = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
        options.enable_shm = enable_shm
        options.is_local = is_local
        options.input_shm_size = input_shm_size
        options.output_shm_size = output_shm_size
        options.kv_cache_dtype = kv_cache_dtype
        options.disable_log_stats = disable_log_stats
        options.enable_sleep_mode = enable_sleep_mode
        self._backend = backend
        if backend == "vlm":
            self.master = VLMMaster(options)
        else:
            self.master = LLMMaster(options)

    def finish(self) -> None:
        try:
            # os.kill(os.getpid(), signal.SIGTERM)
            # os.kill(os.getpid(), signal.SIGKILL)
            utils.terminate_process(os.getpid())
        except Exception:
            pass

    def sleep(self) -> None:
        """Deep sleep: release device HBM (model weights + KV cache) for RL
        training, returning the physical memory to the driver.

        Contents are discarded; after ``wake_up`` weights are re-loaded and
        KV cache is re-prefilled. Requires the engine to be created with
        ``enable_sleep_mode=True``.
        """
        self.master.sleep()

    def wake_up(self, tags: list[str] | None = None) -> None:
        """Re-acquire device HBM previously released by ``sleep``.

        ``tags`` is reserved for future fine-grained wake-up
        (e.g. ["weights"] / ["kv_cache"]); the current version performs a
        full wake-up regardless of ``tags``.
        """
        self.master.wake_up()

    def is_sleeping(self) -> bool:
        return self.master.is_sleeping()

    def update_weights(self, checkpoint_path: str = "") -> None:
        """Reload model weights in place from disk (vllm-ascend ``reload_weights``).

        Typical level-2 RL flow:
            llm.sleep()
            llm.wake_up()                 # re-map empty weight + KV memory
            llm.update_weights("/path/to/new_ckpt")  # reload weights in place
        An empty ``checkpoint_path`` reuses the original model path.
        Requires ``enable_sleep_mode=True``.
        """
        self.master.update_weights(checkpoint_path)

    def generate(
        self,
        prompts: str | list[str] | dict[str, object] | list[dict[str, object]],
        sampling_params: SamplingParams | list[SamplingParams] | None = None,
        wait_for_schedule: bool = True,
        use_tqdm: bool | Callable[..., Any] = True,
        **kwargs: Any,
    ) -> list[RequestOutput]:
        request_params = kwargs.pop("request_params", None)
        if kwargs:
            unknown = ", ".join(kwargs.keys())
            raise TypeError(f"Unexpected keyword arguments: {unknown}")
        if request_params is None:
            request_params = sampling_params
        elif sampling_params is not None:
            raise ValueError("sampling_params and request_params cannot both be set")

        mm_datas = None
        image_urls = None
        if self._is_vllm_style_inputs(prompts):
            from . import mm_utils

            prompts, mm_datas, image_urls = mm_utils.normalize_vllm_style_inputs(prompts)
        else:
            if isinstance(prompts, str):
                prompts = [prompts]
            if not isinstance(prompts, list) or not all(isinstance(x, str) for x in prompts):
                raise TypeError("prompts must be str/list[str] or vLLM-style dicts")

        request_params_list = to_request_params_list(request_params, default_cls=SamplingParams)
        if len(request_params_list) not in (1, len(prompts)):
            raise ValueError("The number of request_params must be 1 or equal to the number of prompts.")

        outputs = [None] * len(prompts)
        progress_bar = None
        progress_bar_lock = threading.Lock()
        tqdm_cls = _get_tqdm(use_tqdm)
        if tqdm_cls is not None:
            progress_bar = tqdm_cls(total=len(prompts), desc="Processed prompts")

        def callback(index: int, output: RequestOutput) -> bool:
            outputs[index] = output
            if progress_bar is not None:
                with progress_bar_lock:
                    progress_bar.update(1)
            return True

        try:
            # schedule all requests
            if self._backend == "vlm":
                if mm_datas is not None:
                    self.master.handle_batch_request(prompts, mm_datas, request_params_list, callback)
                else:
                    if image_urls is None:
                        image_urls = [[] for _ in prompts]
                    self.master.handle_batch_request_with_image_urls(prompts, image_urls, request_params_list, callback)
            else:
                has_images = image_urls is not None and any(image_urls)
                if mm_datas is not None or has_images:
                    raise ValueError("multi_modal_data is only supported for VLM models")
                self.master.handle_batch_request(prompts, request_params_list, callback)

            # TODO: add wait later
            if wait_for_schedule:
                pass

            # generate
            self.master.generate()

            count = len(prompts)
            idx = 0
            while idx < count:
                # wait async output
                if outputs[idx] is None:
                    continue
                if outputs[idx].status is not None and not outputs[idx].status.ok:
                    raise ValidationError(outputs[idx].status.code, outputs[idx].status.message)
                outputs[idx].prompt = prompts[idx]
                idx += 1
        finally:
            if progress_bar is not None:
                progress_bar.close()

        return outputs

    def beam_search(
        self,
        prompts: str | dict[str, str] | list[str | dict[str, str]],
        params: RequestParams | BeamSearchParams | None = None,
        wait_for_schedule: bool = True,
        use_tqdm: bool | Callable[..., Any] = True,
    ) -> list[BeamSearchOutput]:
        if isinstance(prompts, (str, dict)):
            prompts = [prompts]

        parsed_prompts: list[str] = []
        for prompt in prompts:
            if isinstance(prompt, str):
                parsed_prompts.append(prompt)
                continue
            if isinstance(prompt, dict):
                if "prompt" not in prompt:
                    raise ValueError("beam_search prompt dict must contain key 'prompt'")
                parsed_prompts.append(prompt["prompt"])
                continue
            raise TypeError("prompts must be str or dict with key 'prompt'")

        explicit_fields = params.explicit_fields() if isinstance(params, _RequestParamsProxy) else set()
        params = to_request_params(params, default_cls=BeamSearchParams)
        if params.beam_width <= 0:
            raise ValueError("beam_width must be greater than 0")
        elif params.beam_width > 1:
            # Beam search relies on top-k logprob candidates from sampler.
            # Keep this aligned with the LLM request-path default.
            if "logprobs" not in explicit_fields:
                params.logprobs = True
            if "top_logprobs" not in explicit_fields and params.top_logprobs == 0:
                params.top_logprobs = params.beam_width

        outputs = self.generate(
            parsed_prompts, request_params=params, wait_for_schedule=wait_for_schedule, use_tqdm=use_tqdm
        )
        return [BeamSearchOutput(output) for output in outputs]

    def embed(
        self,
        prompts: str | list[str],
        pooling_params: RequestParams | PoolingParams | list[RequestParams | PoolingParams] | None = None,
        wait_for_schedule: bool = True,
        use_tqdm: bool | Callable[..., Any] = True,
    ) -> list[EmbeddingOutput]:
        request_params_list = to_request_params_list(pooling_params, default_cls=PoolingParams)
        for params in request_params_list:
            params.is_embeddings = True

        use_params: RequestParams | list[RequestParams]
        if len(request_params_list) == 1:
            use_params = request_params_list[0]
        else:
            use_params = request_params_list

        outputs = self.generate(
            prompts, request_params=use_params, wait_for_schedule=wait_for_schedule, use_tqdm=use_tqdm
        )
        return [EmbeddingOutput(output) for output in outputs]

    @staticmethod
    def _normalize_selector_values(
        prompts: Sequence[str],
        selector: str | dict | Sequence[str | dict],
    ) -> list[str]:
        def get_literal(value: str | dict) -> str:
            if isinstance(value, str):
                return value
            if isinstance(value, dict):
                selector_type = value.get("type", "literal")
                literal = value.get("value", "")
                if selector_type != "literal":
                    raise ValueError("selector.type must be literal")
                if not isinstance(literal, str) or not literal:
                    raise ValueError("selector.value is required")
                return literal
            raise ValueError("selector must be a string or dict")

        if isinstance(selector, (str, dict)):
            literal = get_literal(selector)
            return [literal for _ in prompts]

        selector_values = list(selector)
        if len(selector_values) != len(prompts):
            raise ValueError("selector count must match prompts count")
        return [get_literal(item) for item in selector_values]

    @staticmethod
    def _build_request_params_list(
        prompts: Sequence[str],
        request_params: RequestParams | Sequence[RequestParams] | None,
    ) -> list[RequestParams]:
        if request_params is None:
            return [RequestParams() for _ in prompts]
        if isinstance(request_params, RequestParams):
            if len(prompts) != 1:
                raise ValueError("request_params must be a list when prompts has multiple items")
            return [request_params]

        params_list = list(request_params)
        if len(params_list) != len(prompts):
            raise ValueError("request_params count must match prompts count")
        return params_list

    def sample(
        self,
        prompts: str | list[str],
        selector: str | dict | Sequence[str | dict],
        request_params: RequestParams | Sequence[RequestParams] | None = None,
        logprobs: int = 5,
        wait_schedule_done: bool = True,
    ) -> list[RequestOutput]:
        if isinstance(prompts, str):
            prompts = [prompts]
        if not prompts:
            return []

        selector_values = self._normalize_selector_values(prompts, selector)
        params_list = self._build_request_params_list(prompts, request_params)
        if len(params_list) > 1:
            # sample() 会原地修改每个 RequestParams（如 request_id/sample_slots）。
            # 若复用同一个对象，会在并发批处理时互相覆盖。
            unique_param_objects = {id(p) for p in params_list}
            if len(unique_param_objects) != len(params_list):
                raise ValueError(
                    "request_params contains duplicated RequestParams objects. "
                    "Please create one RequestParams instance per prompt."
                )

        for i, prompt in enumerate(prompts):
            params = params_list[i]
            if not params.request_id:
                params.request_id = "sample-" + uuid.uuid4().hex

            params.max_tokens = 1
            params.n = 1
            params.best_of = 1
            params.logprobs = True
            params.top_logprobs = logprobs
            params.add_special_tokens = True
            params.is_sample_request = True

            ok, sample_slots = self.master.build_sample_slots(
                params.request_id,
                prompt,
                selector_values[i],
            )
            if not ok:
                raise ValueError("Failed to build sample slots. selector.value must be a stable single special token.")
            params.sample_slots = sample_slots

        outputs = [None] * len(prompts)

        def callback(index: int, output: RequestOutput) -> bool:
            outputs[index] = output
            return True

        self.master.handle_batch_request(prompts, params_list, callback)

        if wait_schedule_done:
            pass

        self.master.generate()

        for i in range(len(outputs)):
            while outputs[i] is None:
                time.sleep(0.01)
            if outputs[i].status is not None and not outputs[i].status.ok:
                raise RuntimeError(f"sample request failed: {outputs[i].status.message}")
            outputs[i].prompt = prompts[i]

        return outputs

File: xllm/cc_api/llm.cpp
/* Copyright 2025 The xLLM Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://github.com/jd-opensource/xllm/blob/main/LICENSE

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "llm.h"

#include <folly/Unit.h>
#include <folly/experimental/coro/Timeout.h>
#include <folly/futures/Future.h>
#include <glog/logging.h>
#include <pthread.h>

#include <atomic>
#include <exception>

#include "internal.h"

namespace xllm {
namespace {
static std::atomic<bool> g_glog_inited = false;
static pthread_mutex_t g_log_init_mutex = PTHREAD_MUTEX_INITIALIZER;

void InitGlog(const std::string& log_dir) {
  pthread_mutex_lock(&g_log_init_mutex);
  if (!g_glog_inited) {
    google::InitGoogleLogging("xllm");
    google::SetLogDestination(google::INFO,
                              (log_dir + "/xllm.log.INFO.").c_str());
    google::SetLogDestination(google::WARNING,
                              (log_dir + "/xllm.log.WARNING.").c_str());
    google::SetLogDestination(google::ERROR,
                              (log_dir + "/xllm.log.ERROR.").c_str());
    google::SetStderrLogging(google::FATAL);

    g_glog_inited = true;
  }
  pthread_mutex_unlock(&g_log_init_mutex);
}
}  // namespace

LLM::LLM() = default;
LLM::~LLM() {
  if (nullptr != llm_core_) {
    delete llm_core_;
    llm_core_ = nullptr;
  }
}

bool LLM::Initialize(const std::string& model_path,
                     const std::string& devices,
                     const XLLM_InitLLMOptions& init_options) {
  if (!init_options.log_dir.empty()) {
    InitGlog(init_options.log_dir);
  }

  if (!std::filesystem::exists(model_path)) {
    LOG(ERROR) << "model path[" << model_path << "] does not exist";
    return false;
  }

  try {
    Options options;
    options.model_path(model_path)
        .task_type(init_options.task)
        .devices(devices)
        .draft_model_path(init_options.draft_model)
        .draft_devices(init_options.draft_devices)
        .backend("llm")
        .block_size(init_options.block_size)
        .max_cache_size(init_options.max_cache_size)
        .max_memory_utilization(init_options.max_memory_utilization)
        .enable_prefix_cache(init_options.enable_prefix_cache)
        .max_tokens_per_batch(init_options.max_tokens_per_batch)
        .max_seqs_per_batch(init_options.max_seqs_per_batch)
        .max_tokens_per_chunk_for_prefill(
            init_options.max_tokens_per_chunk_for_prefill)
        .num_speculative_tokens(init_options.num_speculative_tokens)
        .num_request_handling_threads(init_options.num_request_handling_threads)
        .communication_backend(init_options.communication_backend)
        .rank_tablefile(init_options.rank_tablefile)
        .expert_parallel_degree(init_options.expert_parallel_degree)
        .enable_mla(init_options.enable_mla)
        .enable_chunked_prefill(init_options.enable_chunked_prefill)
        .master_node_addr(init_options.master_node_addr)
        .device_ip(init_options.device_ip)
        .transfer_listen_port(init_options.transfer_listen_port)
        .nnodes(init_options.nnodes)
        .node_rank(init_options.node_rank)
        .dp_size(init_options.dp_size)
        .ep_size(init_options.ep_size)
        .xservice_addr(init_options.xservice_addr)
        .instance_name(init_options.instance_name)
        .enable_disagg_pd(init_options.enable_disagg_pd)
        .enable_schedule_overlap(init_options.enable_schedule_overlap)
        .enable_pd_ooc(init_options.enable_pd_ooc)
        .kv_cache_transfer_mode(init_options.kv_cache_transfer_mode)
        .disable_ttft_profiling(init_options.disable_ttft_profiling)
        .enable_forward_interruption(init_options.enable_forward_interruption)
        .enable_shm(init_options.enable_shm)
        .is_local(init_options.is_local)
        .server_idx(init_options.server_idx);

    llm_core_ = new LLMCore();
    llm_core_->master = std::make_unique<LLMMaster>(options);
    llm_core_->master->run();

    size_t cpu_cores = std::thread::hardware_concurrency();
    size_t thread_num = std::clamp((cpu_cores == 0) ? 8 : cpu_cores / 2,
                                   static_cast<size_t>(4),
                                   static_cast<size_t>(16));
    llm_core_->executor =
        std::make_unique<folly::CPUThreadPoolExecutor>(thread_num);

    std::filesystem::path model_path_fs =
        std::filesystem::path(model_path).lexically_normal();
    std::string model_id;
    if (model_path_fs.has_filename()) {
      model_id = model_path_fs.filename().string();
    } else if (!model_path_fs.empty()) {
      model_id = model_path_fs.string();
    } else {
      model_id = "default";
    }
    llm_core_->model_ids.emplace_back(model_id);

    return true;
  } catch (const std::exception& e) {
    LOG(ERROR) << "LLM initialization failed: " << e.what();
    if (nullptr != llm_core_) {
      delete llm_core_;
      llm_core_ = nullptr;
    }

    return false;
  }
}

XLLM_Response LLM::Completions(const std::string& model_id,
                               const std::string& prompt,
                               uint32_t timeout_ms,
                               const XLLM_RequestParams& request_params) {
  return detail::handle_inference_request(llm_core_,
                                          model_id,
                                          prompt,
                                          timeout_ms,
                                          request_params,
                                          detail::InterfaceType::COMPLETIONS);
}

XLLM_Response LLM::ChatCompletions(
    const std::string& model_id,
    const std::vector<XLLM_ChatMessage>& messages,
    uint32_t timeout_ms,
    const XLLM_RequestParams& request_params) {
  std::vector<Message> internal_messages;
  internal_messages.reserve(messages.size());
  for (const auto& msg : messages) {
    internal_messages.emplace_back(msg.role, msg.content);
  }

  return detail::handle_inference_request(
      llm_core_,
      model_id,
      internal_messages,
      timeout_ms,
      request_params,
      detail::InterfaceType::CHAT_COMPLETIONS);
}
}  // namespace xllm